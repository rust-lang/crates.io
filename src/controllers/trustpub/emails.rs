use crate::email::EmailMessage;
use crates_io_database::models::trustpub::{GitHubConfig, GitLabConfig};
use crates_io_database::models::{Crate, User};

#[derive(Debug, Clone, Copy, serde::Serialize)]
#[serde(tag = "type")]
pub enum ConfigType<'a> {
    GitHub(&'a GitHubConfig),
    GitLab(&'a GitLabConfig),
}

#[derive(serde::Serialize)]
pub struct ConfigCreatedEmail<'a> {
    /// The GitHub login of the email recipient.
    pub recipient: &'a str,
    /// The user who created the trusted publishing configuration.
    pub auth_user: &'a User,
    /// The crate for which the trusted publishing configuration was created.
    pub krate: &'a Crate,
    /// The trusted publishing configuration that was created.
    pub saved_config: ConfigType<'a>,
}

impl ConfigCreatedEmail<'_> {
    pub fn render(&self) -> Result<EmailMessage, minijinja::Error> {
        EmailMessage::from_template("trustpub_config_created", self)
    }
}

#[derive(serde::Serialize)]
pub struct ConfigDeletedEmail<'a> {
    /// The GitHub login of the email recipient.
    pub recipient: &'a str,
    /// The user who deleted the trusted publishing configuration.
    pub auth_user: &'a User,
    /// The crate for which the trusted publishing configuration was deleted.
    pub krate: &'a Crate,
    /// The trusted publishing configuration that was deleted.
    pub config: ConfigType<'a>,
}

impl ConfigDeletedEmail<'_> {
    pub fn render(&self) -> Result<EmailMessage, minijinja::Error> {
        EmailMessage::from_template("trustpub_config_deleted", self)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    use claims::assert_ok;
    use insta::assert_snapshot;

    fn test_user() -> User {
        User {
            id: 1,
            gh_login: "octocat".into(),
            name: Some("The Octocat".into()),
            gh_id: 123,
            gh_avatar: None,
            gh_encrypted_token: vec![],
            account_lock_reason: None,
            account_lock_until: None,
            is_admin: false,
            publish_notifications: true,
        }
    }

    fn test_crate() -> Crate {
        Crate {
            id: 1,
            name: "my-crate".into(),
            updated_at: Utc::now(),
            created_at: Utc::now(),
            description: None,
            homepage: None,
            documentation: None,
            repository: None,
            max_upload_size: None,
            max_features: None,
        }
    }

    fn test_github_config(environment: Option<&str>) -> GitHubConfig {
        GitHubConfig {
            id: 1,
            created_at: Utc::now(),
            crate_id: 1,
            repository_owner_id: 42,
            repository_owner: "rust-lang".into(),
            repository_name: "rust".into(),
            workflow_filename: "publish.yml".into(),
            environment: environment.map(String::from),
        }
    }

    fn test_gitlab_config(environment: Option<&str>) -> GitLabConfig {
        GitLabConfig {
            id: 1,
            created_at: Utc::now(),
            crate_id: 1,
            namespace_id: None,
            namespace: "rust-lang".into(),
            project: "my-crate".into(),
            workflow_filepath: ".gitlab-ci.yml".into(),
            environment: environment.map(String::from),
        }
    }

    #[test]
    fn test_config_created_email() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_with_environment() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitHub(&test_github_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_different_recipient() {
        let email = ConfigCreatedEmail {
            recipient: "team-member",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_gitlab() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitLab(&test_gitlab_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_created_email_gitlab_with_environment() {
        let email = ConfigCreatedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            saved_config: ConfigType::GitLab(&test_gitlab_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration added to my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_with_environment() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitHub(&test_github_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_different_recipient() {
        let email = ConfigDeletedEmail {
            recipient: "team-member",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitHub(&test_github_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_gitlab() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitLab(&test_gitlab_config(None)),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }

    #[test]
    fn test_config_deleted_email_gitlab_with_environment() {
        let email = ConfigDeletedEmail {
            recipient: "octocat",
            auth_user: &test_user(),
            krate: &test_crate(),
            config: ConfigType::GitLab(&test_gitlab_config(Some("production"))),
        };

        let rendered = assert_ok!(email.render());
        assert_snapshot!(rendered.subject, @"crates.io: Trusted Publishing configuration removed from my-crate");
        assert_snapshot!(rendered.body_text);
    }
}
