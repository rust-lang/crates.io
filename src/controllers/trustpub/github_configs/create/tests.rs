use crate::tests::builders::CrateBuilder;
use crate::tests::util::{RequestHelper, Response, TestApp};
use anyhow::anyhow;
use bytes::Bytes;
use crates_io_database::models::token::{CrateScope, EndpointScope};
use crates_io_database::schema::{emails, trustpub_configs_github};
use crates_io_github::{GitHubError, GitHubUser, MockGitHubClient};
use diesel::prelude::*;
use diesel_async::RunQueryDsl;
use insta::{assert_json_snapshot, assert_snapshot};
use serde_json::json;

const URL: &str = "/api/v1/trusted_publishing/github_configs";

const CRATE_NAME: &str = "foo";

fn simple_github_mock() -> MockGitHubClient {
    let mut github_mock = MockGitHubClient::new();
    github_mock.expect_get_user().returning(|login, _| {
        Ok(GitHubUser {
            avatar_url: None,
            email: None,
            id: 42,
            login: login.into(),
            name: None,
        })
    });
    github_mock
}

async fn run_test(payload: impl Into<Bytes>) -> (TestApp, Response<()>) {
    async fn inner(payload: Bytes) -> (TestApp, Response<()>) {
        let (app, _client, cookie_client) = TestApp::full()
            .with_github(simple_github_mock())
            .with_user()
            .await;

        let mut conn = app.db_conn().await;

        CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
            .build(&mut conn)
            .await
            .unwrap();

        (app, cookie_client.post::<()>(URL, payload).await)
    }

    inner(payload.into()).await
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    assert_snapshot!(app.emails_snapshot().await);

    let mut conn = app.db_conn().await;
    let config_ids = trustpub_configs_github::table
        .select(trustpub_configs_github::id)
        .get_results::<i32>(&mut conn)
        .await?;

    assert_eq!(config_ids.len(), 1);
    assert_eq!(config_ids[0], 1);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path_with_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "production",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_body() -> anyhow::Result<()> {
    let (_app, response) = run_test("").await;
    assert_snapshot!(response.status(), @"415 Unsupported Media Type");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Expected request with `Content-Type: application/json`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_empty_json_object() -> anyhow::Result<()> {
    let (_app, response) = run_test("{}").await;
    assert_snapshot!(response.status(), @"422 Unprocessable Entity");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Failed to deserialize the JSON body into the target type: missing field `github_config` at line 1 column 2"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_owner() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "ยง$%&",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository owner name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_repo() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "@foo",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Invalid GitHub repository name"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_workflow() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "workflows/ci.json",
            "environment": null,
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Workflow filename must end with `.yml` or `.yaml`"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_invalid_environment() -> anyhow::Result<()> {
    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": "",
        }
    }))?;

    let (_app, response) = run_test(body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Environment name may not be empty (use `null` to omit)"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unauthenticated() -> anyhow::Result<()> {
    let (app, client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action requires authentication"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_legacy_token_auth() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_token()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_without_trusted_publishing_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from(CRATE_NAME).unwrap()]),
            Some(vec![EndpointScope::PublishUpdate]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wrong_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("other-crate").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this token does not have the required permissions to perform this action"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_token_auth_with_wildcard_crate_scope() -> anyhow::Result<()> {
    let (app, _client, cookie_client, token_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_scoped_token(
            Some(vec![CrateScope::try_from("*").unwrap()]),
            Some(vec![EndpointScope::TrustedPublishing]),
        )
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = token_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_json_snapshot!(response.json(), { ".github_config.created_at" => "[datetime]" });

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_missing_crate() -> anyhow::Result<()> {
    let (_app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"404 Not Found");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"crate `foo` does not exist"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_non_owner() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let other_client = app.db_new_user("other_user").await;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = other_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You are not an owner of this crate"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unknown_github_user() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|login, _| Err(GitHubError::NotFound(anyhow!("User {} not found", login))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Unknown GitHub user or organization"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_github_error() -> anyhow::Result<()> {
    let mut github_mock = MockGitHubClient::new();
    github_mock
        .expect_get_user()
        .returning(|_, _| Err(GitHubError::Other(anyhow!("Internal Server Error"))));

    let (app, _client, cookie_client) = TestApp::full().with_github(github_mock).with_user().await;

    let mut conn = app.db_conn().await;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"500 Internal Server Error");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"Internal Server Error"}]}"#);

    Ok(())
}

#[tokio::test(flavor = "multi_thread")]
async fn test_unverified_email() -> anyhow::Result<()> {
    let (app, _client, cookie_client) = TestApp::full()
        .with_github(simple_github_mock())
        .with_user()
        .await;

    let mut conn = app.db_conn().await;

    diesel::update(emails::table.filter(emails::user_id.eq(cookie_client.as_model().id)))
        .set(emails::verified.eq(false))
        .execute(&mut conn)
        .await?;

    CrateBuilder::new(CRATE_NAME, cookie_client.as_model().id)
        .build(&mut conn)
        .await?;

    let body = serde_json::to_vec(&json!({
        "github_config": {
            "crate": CRATE_NAME,
            "repository_owner": "rust-lang",
            "repository_name": "foo-rs",
            "workflow_filename": "publish.yml",
            "environment": null,
        }
    }))?;

    let response = cookie_client.post::<()>(URL, body).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"You must verify your email address to create a Trusted Publishing config"}]}"#);

    Ok(())
}
