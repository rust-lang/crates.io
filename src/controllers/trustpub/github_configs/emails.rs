use crate::email::Email;

/// Email template for notifying crate owners about a new crate version
/// being published.
#[derive(Debug, Clone)]
pub struct ConfigCreatedEmail<'a> {
    pub recipient: &'a str,
    pub user: &'a str,
    pub krate: &'a str,
    pub repository_owner: &'a str,
    pub repository_name: &'a str,
    pub workflow_filename: &'a str,
    pub environment: &'a str,
}

impl Email for ConfigCreatedEmail<'_> {
    fn subject(&self) -> String {
        let Self { krate, .. } = self;
        format!("crates.io: Trusted Publishing configuration added to {krate}")
    }

    fn body(&self) -> String {
        let Self {
            recipient,
            user,
            krate,
            repository_owner,
            repository_name,
            workflow_filename,
            environment,
        } = self;

        format!(
            "Hello {recipient}!

crates.io user {user} added a new \"Trusted Publishing\" configuration for GitHub Actions to a crate that you manage ({krate}). Trusted publishers act as trusted users and can publish new versions of the crate automatically.

Trusted Publishing configuration:

- Repository owner: {repository_owner}
- Repository name: {repository_name}
- Workflow filename: {workflow_filename}
- Environment: {environment}

If you did not make this change and you think it was made maliciously, you can remove the configuration from the crate via the \"Settings\" tab on the crate's page.

If you are unable to revert the change and need to do so, you can email help@crates.io to communicate with the crates.io support team."
        )
    }
}

/// Email template for notifying crate owners about a Trusted Publishing
/// configuration being deleted.
#[derive(Debug, Clone)]
pub struct ConfigDeletedEmail<'a> {
    pub recipient: &'a str,
    pub user: &'a str,
    pub krate: &'a str,
    pub repository_owner: &'a str,
    pub repository_name: &'a str,
    pub workflow_filename: &'a str,
    pub environment: &'a str,
}

impl Email for ConfigDeletedEmail<'_> {
    fn subject(&self) -> String {
        let Self { krate, .. } = self;
        format!("crates.io: Trusted Publishing configuration removed from {krate}")
    }

    fn body(&self) -> String {
        let Self {
            recipient,
            user,
            krate,
            repository_owner,
            repository_name,
            workflow_filename,
            environment,
        } = self;

        format!(
            "Hello {recipient}!

crates.io user {user} removed a \"Trusted Publishing\" configuration for GitHub Actions from a crate that you manage ({krate}).

Trusted Publishing configuration:

- Repository owner: {repository_owner}
- Repository name: {repository_name}
- Workflow filename: {workflow_filename}
- Environment: {environment}

If you did not make this change and you think it was made maliciously, you can email help@crates.io to communicate with the crates.io support team."
        )
    }
}
