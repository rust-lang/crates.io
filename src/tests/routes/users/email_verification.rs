use super::emails::MockEmailHelper;
use crate::tests::util::{RequestHelper, Response, TestApp};
use insta::assert_snapshot;

pub trait MockEmailVerificationHelper: RequestHelper {
    async fn resend_confirmation(&self, user_id: i32, email_id: i32) -> Response<()> {
        let url = format!("/api/v1/users/{user_id}/emails/{email_id}/resend");
        self.put(&url, &[] as &[u8]).await
    }
}

impl MockEmailVerificationHelper for crate::tests::util::MockCookieUser {}
impl MockEmailVerificationHelper for crate::tests::util::MockAnonymousUser {}

#[tokio::test(flavor = "multi_thread")]
async fn test_no_auth() {
    let (app, anon, user) = TestApp::init().with_user().await;

    let response = anon.resend_confirmation(user.as_model().id, 1).await;
    assert_snapshot!(response.status(), @"403 Forbidden");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"this action requires authentication"}]}"#);

    assert_eq!(app.emails().await.len(), 0);
}

#[tokio::test(flavor = "multi_thread")]
async fn test_wrong_user() {
    let (app, _anon, user) = TestApp::init().with_user().await;
    let user2 = app.db_new_user("bar").await;
    let response = user.resend_confirmation(user2.as_model().id, 1).await;
    assert_snapshot!(response.status(), @"400 Bad Request");
    assert_snapshot!(response.text(), @r#"{"errors":[{"detail":"current user does not match requested user"}]}"#);
    assert_eq!(app.emails().await.len(), 0);
}

#[tokio::test(flavor = "multi_thread")]
async fn test_happy_path() {
    let (app, _anon, user) = TestApp::init().with_user().await;

    // Add an email to the user
    let response = user.add_email(user.as_model().id, "user@example.com").await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_snapshot!(response.text(), @r#"{"id":2,"email":"user@example.com","verified":false,"verification_email_sent":true,"primary":false}"#);

    let response = user
        .resend_confirmation(
            user.as_model().id,
            response.json()["id"].as_u64().unwrap() as i32,
        )
        .await;
    assert_snapshot!(response.status(), @"200 OK");
    assert_snapshot!(response.text(), @r#"{"ok":true}"#);
    assert_snapshot!(app.emails_snapshot().await);
}
