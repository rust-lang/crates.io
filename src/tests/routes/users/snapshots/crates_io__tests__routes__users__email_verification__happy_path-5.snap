---
source: src/tests/routes/users/email_verification.rs
expression: app.emails_snapshot().await
---
To: user@example.com
From: crates.io <noreply@crates.io>
Subject: crates.io: Please confirm your email address
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable


Hello foo!

Welcome to crates.io. Please click the link below to verify your email address:

https://crates.io/confirm/[confirm-token]

Thank you!

--
The crates.io Team
----------------------------------------

To: user@example.com
From: crates.io <noreply@crates.io>
Subject: crates.io: Please confirm your email address
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable


Hello foo!

Welcome to crates.io. Please click the link below to verify your email address:

https://crates.io/confirm/[confirm-token]

Thank you!

--
The crates.io Team
